### ISSUE TITLE
Forbid lru_cache with methods

### ISSUE THREAD

**Issue by orsinium** [wemake-services/wemake-python-styleguide#1770]
# Rule request

## Thesis

+ Forbid using `functools.lru_cache` with regular methods and `properties`. 
+ Still allow usage with `@staticmethod` and `@classmethod`.

Bad:

```python
from functools import lru_cache
class A:
    @lru_cache()
    def b(self):
        return 1

    @property
    @lru_cache()
    def c(self):
        return 1
```

Good:

```python
try:
    from functools import cached_property
except ImportError:
    from cached_property import cached_property

class A:
    def b(self):
        if self._b is None:
             self._b = 1
        return self._b

    @staticmethod
    @lru_cache()
    def also_b():
        return 1

    @cached_property
    def c(self):
        return 1
```

## Reasoning

1. Using it under `@property`, you make unavailable the method for cache invalidation (`cache_clear`).
1. It is slower because it involves, additionally to `__dict__` lookup, search in the cache.
1. It causes a memory leak. `cached_property` stores the data in the object's `__dict__` while `lru_cache` uses shared storage for all objects. So, when you remove an object, the record stays in the memory.
1. It is shared but it is not reused because the cache uses `hash` of the object which is equal to `id` for objects by default which is the object memory address in CPython. So, the result will be calculated and cached for all objects separately, and the cache for removed objects will be never invalidated.

```python
from functools import lru_cache

class A:
    @lru_cache()
    def b(self):
        print(1)
        return None
    c = property(b)

A().c
A().c
A().c

# objects were removed but the cache stays:
A.b.cache_info().currsize
# 3
```

**Comment by sobolevn** [comment]
This should be included in `typed-linter`
